costmap_2d/plugins/shadow_layer.cpp

```diff
@@ -0,0 +1,416 @@
+ *
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2017 6 River Systems
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of 6 River Systems nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Daniel Grieneisen
+ *********************************************************************/
+#include <costmap_2d/shadow_layer.h>
+#include <costmap_2d/costmap_math.h>
+#include <pluginlib/class_list_macros.h>
+
+#include <visualization_msgs/Marker.h>
+
+
+PLUGINLIB_EXPORT_CLASS(costmap_2d::ShadowLayer, costmap_2d::Layer)
+
+using costmap_2d::ObservationBuffer;
+using costmap_2d::Observation;
+
+namespace costmap_2d
+{
+
+void ShadowLayer::onInitialize()
+{
+  ROS_INFO("Initializing shadow layer");
+  ros::NodeHandle nh("~/" + name_), g_nh;
+  rolling_window_ = layered_costmap_->isRolling();
+
+  default_value_ = UNKNOWN;
+
+  ShadowLayer::matchSize();
+  current_ = true;
+
+  global_frame_ = layered_costmap_->getGlobalFrameID();
+
+  dsrv_ = NULL;
+  setupDynamicReconfigure(nh);
+
+  visualization_publisher_ = nh.advertise<visualization_msgs::Marker> ("shadow_points", 1);
+
+}
+
+void ShadowLayer::setupDynamicReconfigure(ros::NodeHandle& nh)
+{
+  dsrv_ = new dynamic_reconfigure::Server<costmap_2d::ShadowPluginConfig>(nh);
+  dynamic_reconfigure::Server<costmap_2d::ShadowPluginConfig>::CallbackType cb = boost::bind(
+      &ShadowLayer::reconfigureCB, this, _1, _2);
+  dsrv_->setCallback(cb);
+}
+
+ShadowLayer::~ShadowLayer()
+{
+  if (dsrv_) 
```

Comment:
the delete operator has a nullptr check internally, so this isn't technically needed.